frame/state-trie-migration/src/lib.rs

```diff
@@ -0,0 +1,1514 @@
+		}
+
+		/// Check if there's any work left, or if we have exhausted the limits already.
+		fn exhausted(&self, limits: MigrationLimits) -> bool {
+			self.last_top.is_none() ||
+				self.dyn_total_items() >= limits.item ||
+				self.dyn_size >= limits.size
+		}
+
+		/// get the total number of keys affected by the current task.
+		pub(crate) fn dyn_total_items(&self) -> u32 {
+			self.dyn_child_items.saturating_add(self.dyn_top_items)
+		}
+
+		/// Migrate keys until either of the given limits are exhausted, or if no more top keys
+		/// exist.
+		///
+		/// Note that this returns after the **first** migration tick that causes exhaustion. In
+		/// other words, this should not be used in any environment where resources are strictly
+		/// bounded (e.g. a parachain), but it is acceptable otherwise (relay chain, offchain
+		/// workers).
+		pub(crate) fn migrate_until_exhaustion(&mut self, limits: MigrationLimits) {
+			log!(debug, "running migrations on top of {:?} until {:?}", self, limits);
+
+			if limits.item.is_zero() || limits.size.is_zero() {
+				// handle this minor edge case, else we would call `migrate_tick` at least once.
+				log!(warn, "limits are zero. stopping");
+				return
+			}
+
+			loop {
+				self.migrate_tick();
+				if self.exhausted(limits) {
+					break
+				}
+			}
+
+			// accumulate dynamic data into the storage items.
+			self.size = self.size.saturating_add(self.dyn_size);
+			self.child_items = self.child_items.saturating_add(self.dyn_child_items);
+			self.top_items = self.top_items.saturating_add(self.dyn_top_items);
+			log!(debug, "finished with {:?}", self);
+		}
+
+		/// Migrate AT MOST ONE KEY. This can be either a top or a child key.
+		///
+		/// This function is the core of this entire pallet.
+		fn migrate_tick(&mut self) {
+			match (self.last_top.as_ref(), self.last_child.as_ref()) {
+				(Some(_), Some(_)) => {
+					// we're in the middle of doing work on a child tree.
+					self.migrate_child();
+				},
+				(Some(ref top_key), None) => {
+					// we have a top key and no child key. 3 possibilities exist:
+					// 1. we continue the top key migrations.
+					// 2. this is the root of a child key, and we start processing child keys (and
+					// should call `migrate_child`).
+					// 3. this is the root of a child key, and we are finishing all child-keys (and
+					// should call `migrate_top`).
+
+					// NOTE: this block is written intentionally to verbosely for easy of
+					// verification.
+					match (
+						top_key.starts_with(DEFAULT_CHILD_STORAGE_KEY_PREFIX),
+						self.prev_tick_child,
+					) {
+						(false, false) => {
+							// continue the top key migration
+							self.migrate_top();
+						},
+						(true, false) => {
+							// start going into a child key.
+							let maybe_first_child_key = {
+								let child_top_key = Pallet::<T>::child_io_key(top_key);
+								sp_io::default_child_storage::next_key(child_top_key, &vec![])
```

Comment:
what if there is a value at the key `&[]` ?
I wrote a quick test, test externalities accept it fine:
```rust
#[test]
fn foo() {
	new_test_ext().execute_with(|| {
		frame_support::storage::child::put(
			&frame_support::storage::child::ChildInfo::new_default(b"abc"),
			&[],
			&3u32,
		);
		assert_eq!(frame_support::storage::child::get(
			&frame_support::storage::child::ChildInfo::new_default(b"abc"),
			&[]
		), Some(3u32));
	});
}
```

Reply:
interesting edge case. Will investigate and make sure it is covered. Note that should we happen such cases, there are manual transactions to detect the issue with.

Reply:
Thiis is a bit hard to replicate. If I add a some data under, for example, `chk1`, with child key `&[]`, then all tests still pass, meaning that this key was also migrated. 

Could it be that calling `next_key(&[])` is enough to potentially migrate it? cc @cheme

Reply:
It should not migrate it, I remember in my branch doing the migration for &[] (even if nothing is in) and calling child next_key only after (there is no way to know if &[] exists but trying to access it).
Was the value you tested with bigger than 32 byte?

Reply:
I was testing with a u32 value. I actually encountered some differences when I change the value to be a vector of 100 bytes. Does it make any difference?

Reply:
Yes, the new trie works the same way as the old one for value size under 33 byte (that is the reason why we can skip writing values smaller than 33 bytes in the migration).
So with u32 value it will not change a thing indeed.

Reply:
damn I should havne known that earlier, half of my tests are useless then, I test values less than 32 bytes 🤦

Reply:
fixed.